### ISSUE TITLE
Graph optimization and other features

### ISSUE THREAD

**Issue by sisp** [tensorflow/tensorflow#3610]
I've been using Theano for about 4 years now and love its flexibility due to the many available low-level ops that allow me to implement complicated and possibly non-standard models without needing to write C++/CUDA code (most of the time). In addition, I can focus almost exclusively on the model design and don't need to think much about numerical stability or suboptimal graph design leading to increased execution times because the graph optimization framework takes care of this. One common example is computing the log-loss of a categorical classifier which I can express naïvely in Theano, but in most (all?) other tools including TensorFlow I need to use an op like ``tf.nn.softmax_cross_entropy_with_logits``. In my opinion, the fact that I need to call these kinds of specialized ops **manually**, i.e. I need to know about them and think about when and how to use them in all kinds of situations, takes away many of the advantages of TensorFlow. Similarly, implementing model optimizers in C++/CUDA directly introduces unnecessary implementation complexity and does not take advantage of the fact that (most likely) all mathematical ops needed to specify the update formulas are already available (CPU and GPU kernels).

I believe TensorFlow has a superior framework design in terms of keeping C++ and Python clearly separated and the inherent multi-device computing capabilities, but the aspects I described above make me very hesitant to use TensorFlow. Are there any plans to add a proper graph optimization framework similar to the one in Theano? And what are your reasons for implementing the model optimizers as individual kernels instead of reusing ops (like using the updates-dictionary in Theano)?

**Comment by poxvoculi** [comment]
Thanks for your comments.  These are interesting issues, however design and feature issues like these are more appropriately addressed on stackoverflow, tagged 'tensorflow'.  This forum is  for bug reports and similar.

**Comment by NeilGirdhar** [comment]
@poxvoculi This is a feature request.  He's suggesting that TensorFlow add a graph optimizer and he provided a motivating example.  Feature requests for TensorFlow belong on Github — not StackOverflow.

**Comment by NeilGirdhar** [comment]
In line with this request, I've been trying to abstract away some differences between theano and tensorflow (so that I can use either one).  My suggested interface for this feature would be something like what follows.  Instead of:

```
input1 = tf.placeholder(tf.float32)
input2 = tf.placeholder(tf.float32)
output = tf.mul(input1, input2)

with tf.Session() as sess:
  print(sess.run([output], feed_dict={input1:[7.], input2:[2.]}))
```

I would like something like:
```
input1 = tf.placeholder(tf.float32)
input2 = tf.placeholder(tf.float32)
output = tf.mul(input1, input2)

# The optimizer runs within tf.compile.
compiled = tf.compile([output], feed_order=[input1, input2])

with tf.Session() as sess:
  # A compiled program is run like a list of bare outputs, but the feeds are provided in the order given at compile time.
  print(sess.run(compiled, [[7.], [2.]]))

  # Alternatively, something like this might be better:
  print(compiled([7.], [2.], session=sess)
```

**Comment by girving** [comment]
We do have plans for further graph optimization, and @NeilGirdhar is correct that design discussions do not belong on StackOverflow.  However, we do not have plans for fixing numerical issues with graph optimization, so I'm going to close the issue for now to keep the issue tracker focused.

**Comment by botev** [comment]
@girving So one thing which is not very clear for instance this - https://github.com/renmengye/tensorflow-forward-ad/issues/2 - a valid question is would tensorflow optimize away the auxiliary variables created. Or in other contexts, would it do operator-fusion. I assume that some of this (especially the second one) is part of XLA, but in general, many of this is part of the graph optimizations. Currently, the documentation talks only about pipelining data and that tensorflow have "many fused Ops" without explicitly listing them. This makes me at least quite uneasy on not knowing what has optimized away and what should I actually bother to write correctly myself.
